src/license.rs

```diff
@@ -70,117 +74,262 @@ impl PoseidonLeaf for DataLeaf {
+        let dec_3 = req
+            .enc_3
+            .decrypt(&k_dh, &req.nonce_3)
+            .expect("decryption should succeed");
 
-        let attr = JubJubScalar::from(112233445566778899u64);
-        let message = sponge::hash(&[npk_user.get_x(), npk_user.get_y(), BlsScalar::from(attr)]);
+        let lpk = JubJubAffine::from_raw_unchecked(dec_1[0], dec_1[1]);
+        let r = JubJubAffine::from_raw_unchecked(dec_2[0], dec_2[1]);
+        let k_lic = JubJubAffine::from_raw_unchecked(dec_3[0], dec_3[1]);
 
-        let sig_lic = Signature::new(&sk_sp, rng, message);
+        let message = sponge::hash(&[lpk.get_x(), lpk.get_y(), BlsScalar::from(*attr)]);
+
+        let sig_lic = Signature::new(&SecretKey::from(ssk_sp.a()), rng, message);
+        let sig_lic_r = JubJubAffine::from(sig_lic.R());
+
+        let nonce_1 = BlsScalar::random(rng);
+        let nonce_2 = BlsScalar::random(rng);
+
+        let enc_1 = PoseidonCipher::encrypt(
+            &[BlsScalar::from(*sig_lic.u()), BlsScalar::from(*attr)],
+            &k_lic,
+            &nonce_1,
+        );
+
+        let enc_2 =
+            PoseidonCipher::encrypt(&[sig_lic_r.get_x(), sig_lic_r.get_y()], &k_lic, &nonce_2);
 
-        let note_type = BlsScalar::from(3u64);
-        let enc = BlsScalar::random(rng);
-        let nonce = BlsScalar::random(rng);
-        let r_user = JubJubAffine::from(GENERATOR_EXTENDED * JubJubScalar::random(rng));
         let pos = BlsScalar::from(1u64);
 
-        // Third, the user computes these values to generate the ZKP later on
+        Self {
+            lsa: StealthAddress::from_raw_unchecked(
+                JubJubExtended::from(r),
+                PublicKey::from_raw_unchecked(JubJubExtended::from(lpk)),
+            ),
+            enc_1,
+            nonce_1,
+            enc_2,
+            nonce_2,
+            pos,
+        }
+    }
+}
+
+#[derive(Default, Debug, Clone, Copy)]
+pub struct LicenseProverParameters {
+    pub lpk: JubJubAffine,   // license public key
+    pub lpk_p: JubJubAffine, // license public key prime
+    pub sig_lic: Signature,  // signature of the license
+
+    pub com_0: BlsScalar,      // Hash commitment 0
+    pub com_1: JubJubExtended, // Pedersen Commitment 1
+    pub com_2: JubJubExtended, // Pedersen Commitment 2
+
+    pub session_hash: BlsScalar,               // hash of the session
+    pub sig_session_hash: dusk_schnorr::Proof, // signature of the ssa_hash
+    pub merkle_proof: PoseidonBranch<DEPTH>,   // Merkle proof for the Proof of Validity
+}
+
+impl LicenseProverParameters {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new<R: RngCore + CryptoRng>(
+        lsa: &StealthAddress,
+        ssk: &SecretSpendKey,
+        lic: &License,
+        psk_sp: &PublicSpendKey,
+        psk_ssp: &PublicSpendKey,
+        k_lic: &JubJubAffine,
+        c: &JubJubScalar,
+        rng: &mut R,
+    ) -> (Self, SessionCookie) {
```

Comment:
It is a bit unexpected to have `new` return `Self` and something extra. I'd rename `new` to a more meaningful name indicating what is being produced.

Reply:
good catch, I'll fix this!